Unresolved symbol Convolution<RGB,RGB>::convolveOneRowDense(int,int) when compiling as static library with Visual Studio

I tried to compile PCL 1.11.1 as a static library in Debug/x64 configuration with the latest Visual Studio 2019 (version 16.7.2). 

I got linker errors for `convolveOneRowDense()` and `convolveOneColDense()`:

    match_linemod_template.obj : error LNK2019: unresolved external symbol "private: struct pcl::RGB __cdecl pcl::filters::Convolution<struct pcl::RGB,struct pcl::RGB>::convolveOneRowDense(int,int)" (?convolveOneRowDense@?$Convolution@URGB@pcl@@U12@@filters@pcl@@AEAA?AURGB@3@HH@Z) referenced in function "private: struct pcl::RGB __cdecl pcl::filters::Convolution<struct pcl::RGB,struct pcl::RGB>::convolveOneRowNonDense(int,int)" (?convolveOneRowNonDense@?$Convolution@URGB@pcl@@U12@@filters@pcl@@AEAA?AURGB@3@HH@Z)

    match_linemod_template.obj : error LNK2019: unresolved external symbol "private: struct pcl::RGB __cdecl pcl::filters::Convolution<struct pcl::RGB,struct pcl::RGB>::convolveOneColDense(int,int)" (?convolveOneColDense@?$Convolution@URGB@pcl@@U12@@filters@pcl@@AEAA?AURGB@3@HH@Z) referenced in function "private: struct pcl::RGB __cdecl pcl::filters::Convolution<struct pcl::RGB,struct pcl::RGB>::convolveOneColNonDense(int,int)" (?convolveOneColNonDense@?$Convolution@URGB@pcl@@U12@@filters@pcl@@AEAA?AURGB@3@HH@Z)

The same errors occur when compiling the current tip of the master branch (commit 050e581b9fff8608c6b251806a27f600a56a4137). So, it looks like #4136 and #4082 did not provide a complete fix for the `convolveOneRowDense()` and `convolveOneColDense()` problem.

To my understanding, things work when PCL is compiled using Visual Studio as a shared library because function 

    template<> pcl::RGB PCL_EXPORTS Convolution<pcl::RGB, pcl::RGB>::convolveOneRowDense (int i, int j);

is declared with `PCL_EXPORTS` in file `convolution.hpp`. Visual Studio handles any DLL-exported inline function by ignoring the `inline` part: the function body is generated and exported from the DLL. The users are forced to call the function as a regular DLL-exported function. No inlining takes place, which, by the way, could hurt performance if inlining was actually intended.

In the static library case, there is no DLL-exporting, and Visual Studio for some reason fails to find the functions.

To make PCL 1.11.1 usable as a static lib on Windows, I patched it by by moving the definitions

    template<> pcl::RGB Convolution<pcl::RGB, pcl::RGB>::convolveOneRowDense(int i, int j)
    template<> pcl::RGB Convolution<pcl::RGB, pcl::RGB>::convolveOneColDense(int i, int j)

from `convolution.cpp` to `convolution.hpp` and removing `PCL_EXPORTS` so as to NOT DLL-export these functions.

--- a/filters/include/pcl/filters/impl/convolution.hpp
+++ b/filters/include/pcl/filters/impl/convolution.hpp
@@ -557,6 +557,9 @@ Convolution<PointIn, PointOut>::convolve_cols_mirror (PointCloudOut& output)
   }
 }
 
+#define PCL_INSTANTIATE_Convolution(Tin, Tout)                                         \
+  template class PCL_EXPORTS Convolution<Tin, Tout>;
+
 } // namespace filters
 } // namespace pcl
 

--- a/filters/src/convolution.cpp
+++ b/filters/src/convolution.cpp
@@ -189,5 +189,17 @@ Convolution<pcl::RGB, pcl::RGB>::convolveOneColDense(int i, int j)
   result.b = static_cast<std::uint8_t>(b);
   return (result);
 }
+
+#ifndef PCL_NO_PRECOMPILE
+#include <pcl/impl/instantiate.hpp>
+#include <pcl/point_types.h>
+
+PCL_INSTANTIATE_PRODUCT(
+    Convolution, ((pcl::RGB))((pcl::RGB)))
+
+PCL_INSTANTIATE_PRODUCT(
+    Convolution, ((pcl::PointXYZRGB))((pcl::PointXYZRGB)))
+#endif // PCL_NO_PRECOMPILE
+
 } // namespace filters
 } // namespace pcl

